certora/specs/ERC20WrapperAdapterReverts.spec

```
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+using Bundler3 as Bundler3;
+
+methods {
+    function _.depositFor(address, uint256) external => DISPATCHER(true);
+    function _.withdrawTo(address, uint256) external => DISPATCHER(true);
+    function _.approve(address, uint256)  external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+    function Bundler3.initiator() external returns address envfree;
+}
+
+// Check that the state didn't change upon a native transfer with 0 ETH using the adapter.
+rule nativeTransferChange(env e, address receiver, uint256 amount) {
+    storage storageBefore = lastStorage;
+    uint256 senderBalanceBefore = nativeBalances[currentContract];
+
+    nativeTransfer(e, receiver, amount);
+
+    // Equivalence is rewritten to avoid issue with quantifiers and polarity.
+    bool noChangeExpectedCondition = amount == max_uint256 && senderBalanceBefore == 0;
+    assert (noChangeExpectedCondition => storageBefore == lastStorage) && (storageBefore == lastStorage => noChangeExpectedCondition);
+}
+
+// Check that the state didn't change upon unwrapping 0 ETH using the adapter.
+rule erc20TransferChange(env e, address token, address receiver, uint256 amount) {
+    uint256 senderBalanceBefore = token.balanceOf(e, currentContract);
+    storage storageBefore = lastStorage;
+
+    erc20Transfer(e, token, receiver, amount);
+
+    // Equivalence is rewritten to avoid issue with quantifiers and polarity.
+    bool noChangeExpectedCondition = amount == 0 || (senderBalanceBefore == 0 && amount == max_uint256);
```

as amount == 0 reverts
```suggestion
    bool noChangeExpectedCondition = amount == max_uint256 && senderBalanceBefore == 0;
```

---

moved to a separate [certora PR](https://github.com/morpho-org/bundler3/pull/245#discussion_r1950627164)